AlpheusCore/DependencyGraph.fs

```diff
@@ -10,13 +10,89 @@ open ItisLab.Alpheus.PathUtils
 open Angara.Data
 open FSharp.Control
 
-let ts = TraceSource("Dependency Graph")
-
 // Disable warning on requiring to override GetHashCode in case of Equals overriding
 // As I want vertices to be alphnumerically sorted, but compared by reference
 #nowarn "0346"
 
 
+type ArtefactLocation  =
+    /// Artefact can be restored from storage
+    |   Remote
+    /// Artefact is currently on disk
+    |   Local
+
+type OutdatedReason =
+    | InputsOutdated
+    | OutputsOutdated
+
+/// computation status of method instance (e.g. scalar or vector element)
+type MethodInstanceStatus =
+    /// Expected outputs version exist
+    | UpToDate of outputs: ArtefactLocation list
+    | Outdated of OutdatedReason
+
+type LinkToArtefactStatus = 
+    /// there is no artefact locally and no expected version on remotes
+    | NotFound
+    /// there is an expected version of artefact on the local disk, and no information about remotes
+    | Local
+    /// there is a version of artefact on the local disk, but it has unexpected version, and no information about remotes
+    | LocalUnexpected 
+    /// there is no any version of artefact on the local disk, but there is a remote artefact of the expected version
+    | Remote
+
+type GroupOfLinksStatus =
+    |   SomeAreLocalUnexpected
+    |   SomeAreNotFound
+    /// All group has expected versions
+    |   AllExist of location:ArtefactLocation list
+
+let getGroupOfLinksStatus checkStoragePresence expectedVersionHashes actualVersionsHashes =
```

Comment:
Checks location of the expected version of the given collection of artefacts and either returns locations of all given artefacts, if they are found; or, otherwise, returns issues.

Reply:
findExpectedArtefacts?

Reply:
Done